src/rust/ide/enso-protocol/src/handler.rs

```diff
@@ -0,0 +1,246 @@
+    logger        : Logger,
+    sender        : Option<UnboundedSender<Event<Notification>>>,
+    ongoing_calls : OngoingCalls<Id,Reply>,
+    #[derivative(Debug="ignore")]
+    processor     : Box<dyn FnMut(TransportEvent) -> Disposition<Id,Reply,Notification>>,
+}
+
+impl <Id,Reply,Notification> HandlerData<Id,Reply,Notification>
+where Id           : Copy + Debug + Display + Hash + Eq + Send + Sync + 'static,
+      Notification : Debug,
+      Reply        : Debug, {
+    fn new<T,P>(transport:T, logger:&Logger, processor:P) -> HandlerData<Id,Reply,Notification>
+    where T : Transport + 'static,
+          P : FnMut(TransportEvent) -> Disposition<Id,Reply,Notification> + 'static {
+        HandlerData {
+            transport     : Box::new(transport),
+            logger        : logger.clone_ref(),
+            sender        : None,
+            ongoing_calls : OngoingCalls::new(logger),
+            processor     : Box::new(processor),
+        }
+    }
+
+    /// Emits event. Clients can consume them through `event_stream`.
+    fn emit_event(&mut self, event:Event<Notification>) {
+        if let Some(sender) = self.sender.as_ref() {
+            // Error can happen if there is no listener. But we don't mind this.
+            let _ = sender.unbounded_send(event);
+        }
+    }
+
+    /// Feeds the reply to complete the corresponding open request.
+    fn process_reply(&mut self, id:Id, reply:Reply) {
+        info!(self.logger,"Processing reply to request {id}: {reply:?}");
+        if let Err(error) = self.ongoing_calls.complete_request(id,reply) {
+            self.emit_error(error);
+        }
+    }
+
+    /// Helper that wraps error into an appropriate event value and emits it.
+    fn emit_error(&mut self, error:impl Into<failure::Error> + Debug) {
+        info!(self.logger,"Emitting error: {error:?}");
+        let event = Event::Error(error.into());
+        self.emit_event(event);
+    }
+
+    /// Handles incoming transport event. The `processor` is used to decide the further processing
+    /// path.
+    ///
+    /// Main entry point for input data while running. Should be connected to the `Transport`s
+    /// output event stream.
+    pub fn process_event(&mut self, event:TransportEvent) {
+        group!(self.logger, "Processing incoming transport event", {
+            debug!(self.logger, "Transport event contents: {event:?}.");
+            let disposition = (self.processor)(event);
+            debug!(self.logger, "Disposition: {disposition:?}");
+            match disposition {
+                Disposition::HandleReply {id,reply} => self.process_reply(id,reply),
+                Disposition::EmitEvent {event} => self.emit_event(event),
+                Disposition::Ignore => {}
+            }
+        });
+    }
+
+    pub fn make_request<F,R>
+    (&mut self, message:&dyn IsRequest<Id=Id>, f:F) -> impl Future<Output=FallibleResult<R>>
+    where F: FnOnce(Reply) -> FallibleResult<R> {
```

Comment:
Can we name it `send_request`? Do it send message immediately or shall I poll the returned future first? It should be covered in docs. And what is `f` here? And why we have &dyn? 

And I was reading this file from top to bottom, and missed the actual docs for this function in `Handler`, so I would put here just a note _See `Handler::make_request` docs_